src/internal/core.rs

```diff
@@ -184,16 +184,14 @@ impl<DP: DependencyProvider> State<DP> {
     fn conflict_resolution(
         &mut self,
         incompatibility: IncompDpId<DP>,
-    ) -> Result<(DP::P, IncompDpId<DP>), PubGrubError<DP>> {
+    ) -> Result<(DP::P, IncompDpId<DP>), NoSolutionError<DP>> {
         let mut current_incompat_id = incompatibility;
         let mut current_incompat_changed = false;
         loop {
             if self.incompatibility_store[current_incompat_id]
                 .is_terminal(&self.root_package, &self.root_version)
             {
-                return Err(PubGrubError::NoSolution(
```

Comment:
Should we go one step further and return only `Err(current_incompat_id)`? This would allow a user to skip the overhead of `build_derivation_tree` if the presence of an error is the only thing they care about.

Reply:
> This would allow a user to skip the overhead of build_derivation_tree if the presence of an error is the only thing they care about.

Sounds good too. This would need to make public the `build_derivation_tree` for those in need. That function is in a core private module. Can it be relocated in a public module? I don’t remember if there is any type or function call in there that requires private access.

Reply:
Ha, it depends on the core `State` because it requires access to the incompat store, which isn’t public right?

Reply:
Our public API would not need to change, because all of this is technically private. `uv` is using it anyway, but they can mark `pub` anything they need to. Making this optimization available to users of our public API, would require some significant thought or we could decide that this is only available to teams using their own resolution loop.

Reply:
Ah I see, you meant just in the context of uv, which already has access to this.

Reply:
For our level of pubgrub usage, we want to call `unit_propagation` ourselves and have a custom report formatter, but we don't want to expose details such as incompatibility ids, so i really like `pub fn unit_propagation(&mut self, package: DP::P) -> Result<(), NoSolutionError<DP>>` as level of abstraction.